tools/apiview/parsers/cpp-api-parser/ApiViewProcessor/ProcessorImpl.cpp

```diff
@@ -0,0 +1,520 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include "ProcessorImpl.hpp"
+#include "ApiViewProcessor.hpp"
+#include <clang/Tooling/CompilationDatabase.h>
+#include <fstream>
+#include <memory>
+#include <nlohmann/json.hpp>
+#include <ostream>
+
+using namespace clang;
+using namespace clang::tooling;
+
+using namespace nlohmann;
+
+inline std::string_view stringFromU8string(std::u8string const& str)
+{
+  return std::string_view(reinterpret_cast<const char*>(str.data()), str.size());
+}
+
+std::vector<std::filesystem::path> GatherSubdirectories(std::filesystem::path const& path)
+{
+  std::vector<std::filesystem::path> subdirectories;
+  for (auto& entry : std::filesystem::directory_iterator(path))
+  {
+    if (entry.is_directory())
+    {
+      subdirectories.push_back(entry.path());
+      auto inner = GatherSubdirectories(entry.path());
+      subdirectories.insert(subdirectories.end(), inner.begin(), inner.end());
+    }
+  }
+  return subdirectories;
+}
+std::string replaceAll(
+    std::string_view const& source,
+    std::string_view const& oldValue,
+    std::string_view const& newValue)
+{
+  std::string newString;
+  newString.reserve(source.size());
+  size_t findPos{};
+  size_t lastPos{};
+
+  while (std::string::npos != (findPos = source.find(oldValue, lastPos)))
+  {
+    newString.append(source, lastPos, findPos - lastPos);
+    newString += newValue;
+    lastPos = findPos + oldValue.length();
+  }
+  newString += source.substr(lastPos);
+  return newString;
+}
+
+const nlohmann::json JsonFromConfigurationPath(
+    std::string_view configurationRoot,
+    std::string_view const& configurationFileName)
+{
+  std::filesystem::path configurationFilePath{configurationRoot};
+  configurationFilePath /= configurationFileName;
+  std::ifstream configurationFile{configurationFilePath};
+  if (!configurationFile.is_open())
+  {
+    throw std::runtime_error(
+        "Unable to open configuration file: " + configurationFilePath.string());
+  }
+  nlohmann::json configurationJson;
+  configurationFile >> configurationJson;
+  return configurationJson;
+}
+
+ApiViewProcessorImpl::ApiViewProcessorImpl(
+    std::string_view directoryToProcess,
+    std::string_view const& configurationFileName)
+    : ApiViewProcessorImpl(
+        directoryToProcess,
+        JsonFromConfigurationPath(directoryToProcess, configurationFileName))
+{
+}
+
+ApiViewProcessorImpl::ApiViewProcessorImpl(
+    std::string_view directoryToProcess,
+    nlohmann::json const& configurationJson)
+    : m_currentSourceRoot{std::filesystem::absolute(directoryToProcess)},
+      m_classDatabase{std::make_unique<AzureClassesDatabase>(this)}
+{
+  // CHDIR to the directory to process so relative paths in the configuration are properly resolved.
+  auto currentDirectory{std::filesystem::current_path()};
```

Comment:
Do you want to put this into a holder such that no matter what happens, when this function exits we reset the cwd

Reply:
Done - it also cleans up the ProcessApiView tool.